### PR TITLE
chore: replace transfer operation with input and output

### PR DESCRIPTION
**Pull request by rhyslbw**
# Description
Currently the only `operationType` is `transfer`, which for Cardano, will be insufficient to properly describe the source and destination of fund movement. This issue is prompted by considering how reward withdrawals, key deposits, and reclaims, will be defined.

# Proposed Solution
Prepares for the Cardano-specific transaction operation types by aligning the standard UTxO operations with the rosetta-bitcoin
implementation.

# Testing

Testing is performed offline until the `check:construction` is automated using a pre-funded account. Please note, the fund return is not working, however is non-critical.

## check:construction
```
Success: {"create_account":2,"transfer":2}

+--------------------------+--------------------------------+-------+
| CHECK:CONSTRUCTION STATS |          DESCRIPTION           | VALUE |
+--------------------------+--------------------------------+-------+
| Addresses Created        | # of addresses created         |     8 |
+--------------------------+--------------------------------+-------+
| Transactions Created     | # of transactions created      |     3 |
+--------------------------+--------------------------------+-------+
| Stale Broadcasts         | # of broadcasts missing after  |     0 |
|                          | stale depth                    |       |
+--------------------------+--------------------------------+-------+
| Transactions Confirmed   | # of transactions seen         |     3 |
|                          | on-chain                       |       |
+--------------------------+--------------------------------+-------+
| Failed Broadcasts        | # of transactions that         |     0 |
|                          | exceeded broadcast limit       |       |
+--------------------------+--------------------------------+-------+
+------------------------------+-------+
| CHECK:CONSTRUCTION WORKFLOWS | COUNT |
+------------------------------+-------+
| request_funds                |     1 |
+------------------------------+-------+
| create_account               |     8 |
+------------------------------+-------+
| transfer                     |     3 |
+------------------------------+-------+
| return_funds                 |     0 |
+------------------------------+-------+

badger 2020/10/14 01:13:18 INFO: Storing value log head: {Fid:0 Len:30 Offset:51247}
```
## check:data byron_sample
```
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |       10000 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |          13 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |          36 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |          34 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |         244 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+
```

## check:data shelley_sample
```
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |        5001 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |        6427 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |       28022 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |       23262 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |       18233 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+
```

## check:data shelley_transition
```
Success: Index End Condition [Index: 4495000]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |        5001 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |       21989 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |      134756 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |      100834 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |       32858 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+

```

### CODE REVIEW

**Review by AlanVerbner — APPROVED** [review]
:rocket:

**Review by AlanVerbner — DISMISSED** [review]
lgtm

File: test/check/configuration/construction/workflow.ros [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      "minimum_balance":{
+        "value": {{reserved_amount}},
+        "currency": {{currency}}
+      }
+    });
+    recipient_amount = {{sender.balance.value}} - {{max_fee_amount}};
+    print_message({"recipient_amount": {{recipient_amount}}});
+    sender_amount = 0 - {{sender.balance.value}};
+    transfer_dry_run.confirmation_depth = 1;
+    return_address = load_env("FUND_RETURN_ADDRESS");
```
Comment (rhyslbw) [inline test/check/configuration/construction/workflow.ros:177]: This isn't working [as expected](https://github.com/coinbase/rosetta-sdk-go/blob/0de1c247d390bedaeda5f1ac8b6923b975f32392/constructor/dsl/README.md#basic-example), so is a placeholder at the moment

File: test/check/configuration/construction/workflow.ros [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    confirmation_depth = 1;
+    suggested_fee = find_currency_amount({
+      "currency": {{currency}},
+      "amounts": {{transfer_dry_run.suggested_fee}}
+    });
+    recipient_amount = {{sender.balance.value}} - {{suggested_fee.value}};
+    dust_amount = 600;
+    recipient_minus_dust = {{recipient_amount}} - {{dust_amount}};
+    assert({{recipient_minus_dust}});
+    operations = [
```
Comment (rhyslbw) [inline test/check/configuration/construction/workflow.ros:218]: This is also identical to the `transfer_dry_run.operations`, but I can't work out how to reference it.

File: test/check/configuration/construction/configuration.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "construction": {
+    "block_broadcast_limit": 0,
+    "broadcast_behind_tip": false,
+    "broadcast_limit": 0,
+    "clear_broadcasts": false,
+    "constructor_dsl_file": "workflow.ros",
```
Comment (AlanVerbner) [inline test/check/configuration/construction/configuration.json:7]: I'm not sure if this is needed

File: test/check/configuration/construction/workflow.ros [hunk 4]
```diff
@@ -0,0 +1,244 @@
+request_funds(1){
```
[the remark this answers is not in the document]
Comment (rhyslbw) [reply test/check/configuration/construction/workflow.ros:1]: It was included for visibility, but I'll remove it, and add to another branch that can remain WIP

File: test/check/configuration/construction/configuration.json (hunk 3 above)
Comment (rhyslbw) [reply test/check/configuration/construction/configuration.json:7]: This whole config file will be moved to a WIP branch

**Pull request merged**